### ISSUE TITLE
Saved Variables: Add TOC field to load saved variables before other files in addons

### ISSUE THREAD

**Issue by Meorawr** [Stanzilla/WoWUIBugs#414]
The existing behaviour of loading saved variables after all other files in an addon and adds a bit of unnecessary complexity and isn't entirely intuitive to newer authors. This feature request proposes a new *opt-in* TOC field that would re-sequence the loading process such that saved variables are loaded before all other files in the TOC.

# Proposal

The proposal is to add a new boolean TOC metadata field - tentatively named "SavedVariablesLoadFirst" - which if set would cause all saved variables files for the addon to be loaded *before* the first file listed in the TOC. This field would be **disabled by default** to maintain compatibility with existing addons.

```
## Title: My really cool addon!
## SavedVariables: MyAddOn_DB
## SavedVariablesLoadFirst: 1

File1.lua
File2.lua
```

The `ADDON_LOADED` and `SAVED_VARIABLES_TOO_LARGE` events would continue to be sequenced as they are today, effectively firing *after* the final file specified in the TOC has been loaded.

# Considerations

## What benefits does this bring?

**For most addons and scripts** you'll no longer need to make a frame and register for an event or use [EventUtil.ContinueOnAddOnLoaded](https://github.com/Gethe/wow-ui-source/blob/8cc5f29bcd45dcbe3ed293cfb146b0178b2d4e92/Interface/SharedXML/EventUtil.lua#L52) just to be notified of when your saved variables are loaded - they'll just be in the environment from the get-go for you to use directly in script bodies.

**For newer authors** saved variables management requires additionally learning how to set up a frame and process the event. Odds are they'll have to do this at some point anyway, but the existing behaviour appears to often be counter to their expectations of how saved variables work and causes some confusion. Teaching newer authors about this flag should hopefully be easier than having to explain anonymous frames and event handling, even if it isn't the default.

**For initialization tasks that are dependent upon saved variables** you may be able to cut down on some of the complexity and do the work directly in script bodies. An example might be addons that provide options to override their configured locale, for which any initialization dependent upon translation strings - like creating UI elements or static data tables - needs to be deferred until the saved variables are loaded.

## What impact does this have on existing addons?

As-proposed, none. The default behaviour is for this flag to be off which will retain today's behaviour of loading saved variables after all script files in the addon.

## Why not make this the default?

It's reasonably likely that there's some practices in a small amount of existing addons that may be incompatible with changing the default behaviour.

For example, an author may be initializing their saved variables global in a script body (`MyAddOn_DB = { <defaults> }`)  with the expectation that if this is the first time a user has logged in with the addon installed then the defaults wouldn't get replaced, and in effect become the saved configuration on logout.

In this case were the default behaviour to change then such an addon would end up replacing its saved data with defaults on every load. It may not be particularly easy for an author to track down this hidden change of defaults unless sufficiently communicated ahead of time.

## Could the default change in an expansion patch?

Probably - the only concern would be communicating this change sufficiently well enough for authors to either opt-out of it, or make changes to ensure they won't break. As expansion patches often have larger breaking changes anyway, such a border feels like it would be the safest possible point at which the default behaviour could change.

**Comment by Ska-Doosh** [comment]
Or just have two new .toc fields:
```
## PreSavedVariables: MyAddOn_DB2
## PreSavedVariablesPerCharacter: MyAddOn_CharDB2
```
those saved variables would be loaded before addon while normal saved variables would be loaded after addon, new authors could be directed to the new way while old authors can switch to the new system when they are ready. Also those variables would be saved in different file like _pre-Addon_Name.lua_ and it would give addons with longer saved variables a chance to split it into two files rather than two addons.

**Comment by Meorawr** [comment]
That specific idea has a rather large problem in that changing a `SavedVariables` line to a `PreSavedVariables` line would result in you losing all your data, as the client would try to load it from the initially-non-existant "Pre" file instead.

You could specify _both_ lines but you're going to then have to keep a duplicate of all your data on-disk for a potentially very long time as not all users will update with any reasonable timeframe.

> it would give addons with longer saved variables a chance to split it into two files rather than two addons.

As for that problem, that's not really the sort of thing to kick down the road with the bandaid of "more files". If an addon is running into the constant limit in its saved variables it really should just be serializing some of the bulk data into strings.

**Comment by Ska-Doosh** [comment]
I do not think that data would be lost, it just tells the client where to save it and how to load it. But both files are always loaded before they are saved, so if you want to transfer you can simply set ##SavedVariables: MyAddonVariableThatWillNeverBeSet

**Comment by funkydude** [comment]
I would far rather use the opportunity to implement the SavedVariables privately, instead of just changing how the current SavedVariables load.

You can have the TOC directive be named whatever you want, but the table should be part of the addon scope `...` which would obviously be available from the very first line of the addon, accomplishing what you originally wanted, and giving devs the choice to make their settings public or not.

Transferring the settings from an old global settings table to the new private table is not a big deal, you already discussed it above.

**Comment by DFortun81** [comment]
I second this idea. It would be nice to not have to wait until the VARIABLES_LOADED event before being allowed to do anything with your saved variables.

**Comment by urnati** [comment]
Defaulting to private under the addon would be great. Below could signify both - load first and where to place.

`## SavedVariablesScope: MyAddOn`

Per the above example there would then be a table MyAddOn.MyAddOn_DB with data or {} as default. Fully available to File1.lua which, if the table is {}, populate with the addon defaults.

**Comment by Meorawr** [comment]
This appears to be resolved in 11.1.5.59571 with the addition of a `LoadSavedVariablesFirst` directive.
